src/common/json.cc

```diff
@@ -527,15 +525,118 @@ Json JsonReader::ParseObject() {
 }
 
 Json JsonReader::ParseNumber() {
-  std::string substr = raw_str_.substr(cursor_.Pos(), kMaxNumLength);
-  size_t pos = 0;
+  // Adopted from sajson with some simplifications and small optimizations.
+  char const* p = raw_str_.c_str() + cursor_.Pos();
+  char const* const beg = p;  // keep track of current pointer
+
+  // TODO(trivialfis): Add back all the checks for number
+  bool negative = false;
+  if ('-' == *p) {
+    ++p;
+    negative = true;
+  }
+
+  bool is_float = false;
+
+  using ExpInt = std::remove_const<
+    decltype(std::numeric_limits<Number::Float>::max_exponent)>::type;
+  constexpr auto kExpMax = std::numeric_limits<ExpInt>::max();
+  constexpr auto kExpMin = std::numeric_limits<ExpInt>::min();
+
+  JsonInteger::Int i = 0;
+  double f = 0.0;  // Use double to maintain accuracy
+
+  if (*p == '0') {
+    ++p;
+  } else {
+    char c = *p;
+    do {
+      ++p;
+      char digit = c - '0';
+      i = 10 * i + digit;
+      c = *p;
+    } while (std::isdigit(c));
+  }
+
+  ExpInt exponent = 0;
+  const char *const dot_position = p;
+  if ('.' == *p) {
+    is_float = true;
+    f = i;
+    ++p;
+    char c = *p;
+
+    do {
+      ++p;
+      f = f * 10 + (c - '0');
+      c = *p;
+    } while (std::isdigit(c));
+  }
+  if (is_float) {
+    exponent = dot_position - p + 1;
+  }
+
+  char e = *p;
+  if ('e' == e || 'E' == e) {
+    if (!is_float) {
+      is_float = true;
+      f = i;
+    }
+    ++p;
+
+    bool negative_exponent = false;
+    if ('-' == *p) {
+      negative_exponent = true;
+      ++p;
+    } else if ('+' == *p) {
+      ++p;
+    }
 
-  Number::Float number{0};
-  number = std::stof(substr, &pos);
-  for (size_t i = 0; i < pos; ++i) {
-    GetNextChar();
+    ExpInt exp = 0;
+
+    char c = *p;
+    while (std::isdigit(c)) {
+      unsigned char digit = c - '0';
+      if (XGBOOST_EXPECT(exp > (kExpMax - digit) / 10, false)) {
+        CHECK_GT(exp, (kExpMax - digit) / 10) << "Overflow";
+      }
+      exp = 10 * exp + digit;
+      ++p;
+      c = *p;
+    }
+    static_assert(-kExpMax >= kExpMin, "exp can be negated without loss or UB");
+    exponent += (negative_exponent ? -exp : exp);
+  }
+
+  if (exponent) {
+    CHECK(is_float);
+    // If d is zero but the exponent is huge, don't
+    // multiply zero by inf which gives nan.
+    if (f != 0.0) {
+      // Only use exp10 from libc on gcc+linux
+#if !defined(__GNUC__) || defined(_WIN32) || defined(__APPLE__)
+#define exp10(val) std::pow(10, (val))
+#endif  // !defined(__GNUC__) || defined(_WIN32) || defined(__APPLE__)
+      f *= exp10(exponent);
```

Comment:
Is `exp10` faster?

Reply:
@hcho3 Sorry I never benchmark this particular function knowing that the newly revised `ParseNumber` is no longer a bottle neck.